src/librustc/mir/interpret/memory.rs

```diff
@@ -247,110 +184,101 @@ impl<'tcx> MemoryPointer {
 // Top-level interpreter memory
 ////////////////////////////////////////////////////////////////////////////////
 
-pub struct Memory<'a, 'tcx, M: Machine<'tcx>> {
+pub struct Memory<'a, 'tcx: 'a, M: Machine<'tcx>> {
     /// Additional data required by the Machine
     pub data: M::MemoryData,
 
+    /// Helps guarantee that stack allocations aren't deallocated via `rust_deallocate`
+    alloc_kind: HashMap<u64, MemoryKind<M::MemoryKinds>>,
+
     /// Actual memory allocations (arbitrary bytes, may contain pointers into other allocations).
-    alloc_map: HashMap<u64, Allocation<'tcx, M::MemoryKinds>>,
+    alloc_map: HashMap<u64, Allocation>,
 
-    /// The AllocId to assign to the next new regular allocation. Always incremented, never gets smaller.
-    next_alloc_id: u64,
+    /// Actual memory allocations (arbitrary bytes, may contain pointers into other allocations).
+    ///
+    /// Stores statics while they are being processed, before they are interned and thus frozen
+    uninitialized_statics: HashMap<u64, Allocation>,
 
     /// Number of virtual bytes allocated.
     memory_usage: u64,
 
     /// Maximum number of virtual bytes that may be allocated.
     memory_size: u64,
 
-    /// Function "allocations". They exist solely so pointers have something to point to, and
-    /// we can figure out what they point to.
-    functions: Vec<Instance<'tcx>>,
-
-    /// Inverse map of `functions` so we don't allocate a new pointer every time we need one
-    function_alloc_cache: HashMap<Instance<'tcx>, AllocId>,
-
-    /// Target machine data layout to emulate.
-    pub layout: &'a TargetDataLayout,
-
-    /// A cache for basic byte allocations keyed by their contents. This is used to deduplicate
-    /// allocations for string and bytestring literals.
-    literal_alloc_cache: HashMap<Vec<u8>, AllocId>,
-
     /// To avoid having to pass flags to every single memory access, we have some global state saying whether
     /// alignment checking is currently enforced for read and/or write accesses.
     reads_are_aligned: Cell<bool>,
     writes_are_aligned: Cell<bool>,
 
     /// The current stack frame.  Used to check accesses against locks.
     pub(super) cur_frame: usize,
+
+    pub tcx: TyCtxt<'a, 'tcx, 'tcx>,
+
+    /// Memory regions that are locked by some function
+    locks: HashMap<u64, RangeMap<LockInfo<'tcx>>>,
```

Comment:
Why did locks move out of the allocation?

Nested maps don't make much sense, shouldn't they become just one large map then? But I moved away from that due to performance concerns.

Reply:
> Why did locks move out of the allocation?

We can't intern RangeMaps in the compiler.

> Nested maps don't make much sense

RangeMap is over byte ranges, they don't include the alloc ids. We could add them, but I don't really see an issue with this. We can optimize them later if we really need it.

Reply:
Okay.